designs/NuGet-ClientCertificates.md

```diff
@@ -0,0 +1,200 @@
+
+All .NET Core customers.
+
+## Requirements
+
+* Ability to specify client certificates from common [certificate stores](https://docs.microsoft.com/en-us/dotnet/framework/wcf/feature-details/working-with-certificates#certificate-stores)
+* Ability to specify client certificates from external files in standard [formats](https://en.wikipedia.org/wiki/X.509#Certificate_filename_extensions)
+* Ability to configure settings with command line
+
+## Solution
+
+1) Find specified in NuGet configuration file certificates (common implementation for [all products](https://github.com/NuGet/NuGet.Client/tree/dev/src/NuGet.Core/NuGet.Configuration) )
+2) Apply them prior any web request to appropriate http handler (product specific)
+
+## NuGet configuration changes
+
+Provide new configuration section `clientCertificates` which may have children of 2 types:
+
+1. `fromStorage` - certificate import from [certificate store](https://docs.microsoft.com/en-us/dotnet/framework/wcf/feature-details/working-with-certificates#certificate-stores). Internally uses [X509Certificate2Collection.Find](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.x509certificate2collection.find?view=netframework-4.8#System_Security_Cryptography_X509Certificates_X509Certificate2Collection_Find_System_Security_Cryptography_X509Certificates_X509FindType_System_Object_System_Boolean_) method.
+Can be configured with:
+    - Attribute `name`. Required. Unique setting identifier.
+    - Child item `<Add Key="StoreLocation" Value="[values]" />`. [Possible values](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.storelocation?view=netframework-4.8#fields). Optional. Equals `CurrentUser` by default.
+    - Child item `<Add Key="StoreName" Value="[values]" />`. [Possible values](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.storename?view=netframework-4.8#fields). Optional. Equals `My` by default.
+    - Child item `<Add Key="FindType" Value="[values]" />`. [Possible values](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.x509findtype?view=netframework-4.8#fields). Optional. Equals `FindByThumbprint` by default.
+    - Child item `<Add Key="FindValue" Value="value" />`. `value` - the search criteria as an object. Required.
+
+2. `fromFile` - certificate import from file (DER encoded x.509, Base-64 encoded x.509, PKCS)
+Can be configured with:
+    - Attribute `name`. Required. Unique setting identifier.
+    - Child item `<Add Key="Path" Value="value" />`. `value` - Absolute or relative path to certificate file. Required.
+    - Child item `<Add Key="Password" Value="value" />`. `value` - Plain or encrypted password string. Optional. Encrypted in same manner as PackageSourceCredential password.
+
+## Configuration example
+
+```xml
+<configuration>
+...
+    <clientCertificates>	
+        <fromStorage name="unique name 1">
```

Comment:
`<fromStore>`? Storage is rather ambiguous (web service guy here thought Azure Storage first, haha). Store matches the .NET features a little closer, e.g. `X509Store`.

Reply:
At linux there are no such thing as X509Store. But still they have bunch of [places](https://www.getpagespeed.com/server-setup/ssl-directory).

Reply:
I think I agree with Joel here. `Store` is specific enough for .NET folks and at the same time as abstract as `storage` so should work for linux as well.